Add einsum

fmayo96 opened the pull request:
**Description**
Added a einsum, a numpy.einsum (Evaluates the Einstein summation convention on the operands.) version for Qobj dimensions.

**Related issues or PRs**
fixes #2340

** Question: Not sure if I should update the documentation and how. Just in case I did not modify the docs yet.

Ericgig commented on qutip/core/dimensions.py:
```diff
@@ -352,6 +352,31 @@ def _frozen(*args, **kwargs):
+    result = np.einsum(subscripts, *operands_array)
+    if result.shape == ():
+        return result
+    dims = [[d] for d in result.shape]
```
Not sure if this work well with compound dimensions.
Could you add tests with multiple qubits systems.

Ericgig replied:
Sometime like 
`dims = [[d for d in result.shape[:result.ndim // 2], [d for d in result.shape[result.ndim // 2:]]`
would keep the compound dimensions.
We should raise warning or error if the output ndim is odd.

Ericgig reviewed (commented):
Thank you for tackling this issue.
Other than confirming the dims are good in multiple qubits systems, it looks good.

hodgestar commented:
Thanks @fmayo96.

Question: Should the `einsum` function be in `dimensions.py`? And should it be internal or exposed via the public API? It looks generically useful but what specific use cases do we have in mind?

BoxiLi commented:
The simplest use case is the circuit simulation, where e.g. single qubit unitaries are multiplied to multiqubit states. For Hamiltonian dynamics I cannot immediately think of an application cause usually you need to sum many terms. Unless we develop a troterrizatiom solver.

fmayo96 commented:
I have corrected the code for compound systems and added tests for two and three qubits. I am not sure about which is the best place for this function, I made it to solve the Issue #2340. If you consider it fits better in a different place I will be happy to modify it

Ericgig reviewed (commented):
Sorry, but the first attempt was better.
We want the `einsum` on the dimensions, not the shape. 
With a 2 qubits operator, there should be 4 indices, 6 indices for 3 qubits, etc.
The output for compound system should also be compound.
We want support for operations that could not be done with any mix of `matmul`, `trace`, etc.

Here are some example of use:

```
einsum("ikjl,jm->ikml", [tensor(sigmaz(), sigmaz()), sigmaz()]) == tensor(qeye(2), sigmaz())
einsum("ijkl->kjil", [tensor(sigmam(), sigmaz())]) == tensor(sigmap(), sigmaz())
```

There are no place for it, a new file (`qutip/core/einsum.py`) would be good, but it's fine to keep in `dimensions.py` for now. We will move it later.

fmayo96 commented:
Oh, I am sorry, I have misunderstood what you wanted. I think that now I have fixed it. Thank you very much for the feedback!

Ericgig commented on qutip/core/dimensions.py:
```diff
@@ -352,6 +352,32 @@ def _frozen(*args, **kwargs):
+    result = np.einsum(subscripts, *operands_array)
+    if result.shape == ():
+        return result
+    dims = [d for d in result.shape[:result.ndim // 2]],
+    [[d for d in result.shape[result.ndim // 2:]]]
```
```suggestion
    dims = [
        [d for d in result.shape[:result.ndim // 2]],
        [d for d in result.shape[result.ndim // 2:]]
    ]
```

Ericgig reviewed (approved):
Thank you
